exp/internal/controllers/machinepool_controller_noderef.go

```diff
@@ -94,30 +95,10 @@ func (r *MachinePoolReconciler) reconcileNodeRefs(ctx context.Context, cluster *
 	log.Info("Set MachinePools's NodeRefs", "noderefs", mp.Status.NodeRefs)
 	r.recorder.Event(mp, corev1.EventTypeNormal, "SuccessfulSetNodeRefs", fmt.Sprintf("%+v", mp.Status.NodeRefs))
 
-	// Reconcile node annotations.
-	for _, nodeRef := range nodeRefsResult.references {
-		node := &corev1.Node{}
-		if err := clusterClient.Get(ctx, client.ObjectKey{Name: nodeRef.Name}, node); err != nil {
-			log.V(2).Info("Failed to get Node, skipping setting annotations", "err", err, "nodeRef.Name", nodeRef.Name)
-			continue
-		}
-		patchHelper, err := patch.NewHelper(node, clusterClient)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		desired := map[string]string{
-			clusterv1.ClusterNameAnnotation:      mp.Spec.ClusterName,
-			clusterv1.ClusterNamespaceAnnotation: mp.GetNamespace(),
-			clusterv1.OwnerKindAnnotation:        mp.Kind,
-			clusterv1.OwnerNameAnnotation:        mp.Name,
-		}
-		// Add annotations and drop NodeUninitializedTaint.
-		if annotations.AddAnnotations(node, desired) || taints.RemoveNodeTaint(node, clusterv1.NodeUninitializedTaint) {
-			if err := patchHelper.Patch(ctx, node); err != nil {
-				log.V(2).Info("Failed patch node to set annotations and drop taints", "err", err, "node name", node.Name)
-				return ctrl.Result{}, err
-			}
-		}
+	// Reconcile node annotations and taints.
+	err = r.patchNodes(ctx, clusterClient, nodeRefsResult.references, mp)
```

Comment:
Just wondering: does whether or not we needed to patch the nodes suggest how much later we should requeue below, if replicas and the length of noderefs is not at the terminal state?

Reply:
No, that gets decided later https://github.com/kubernetes-sigs/cluster-api/pull/8462/files#diff-8b643e9dff0f3fa8e0d24ae8b7dc2c91739400fb73344b804dcf03994dc4e1eeR107

Note this is only returning if `err != nil`, which is equivalent to before (old line 106 and 118)